jax/interpreters/numpy_eval.py

```diff
@@ -0,0 +1,603 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from contextlib import contextmanager
+from typing import Any, Dict, Callable
+import builtins
+import itertools
+import numpy as np
+import opt_einsum
+import scipy.special
+
+from .. import config, core, dtypes, curry, lax
+from ..interpreters import xla
+
+_slice = builtins.slice
+_min = builtins.min
+_map = builtins.map
+
+np_impl: Dict[core.Primitive, Callable[..., Any]] = {}
+
+class NumpyEvalTrace(core.Trace):
+  def pure(self, x): return x
+  lift = sublift = pure
+
+  def process_primitive(self, primitive, tracers, params):
+    impl = np_impl.get(primitive)
+    if impl is None:
+      raise NotImplementedError(f"No original numpy implementation for "
+                                f"'{primitive}' primitive found.")
+    return impl(*tracers, **params)
+
+  def process_call(self, primitive, f, tracers, params):
+    return f.call_wrapped(*tracers)
+  process_map = process_call
+
+@contextmanager
+def numpy_eval():
+  """
+  Makes JAX code evaluate with NumPy instead of XLA.
+
+  >>> @numpy_eval()
+  ... def some_fun(...):
+  ...  # jax code
+
+  Inline:
+
+  >>> from jax.interpreters.numpy_eval import numpy_eval
+  ... with numpy_eval():
+  ...   # jax code
+
+  ``jit(numpy_eval()(some_fun))`` will collapse constant calculations using NumPy,
+  so that no operations are performed on the XLA device during compilation.
+
+  Using `numpy_eval` is thread-safe.
+  """
+  assert config.omnistaging_enabled, \
+    "The NumPy backend requires omnistaging. Set flag JAX_OMNISTAGING=True or " \
+    "call jax.config.enable_omnistaging() before using jax."
+  with core.new_base_master(NumpyEvalTrace):
+    yield
+
+neg = np.negative
```

Comment:
I wonder if it would be preferable to keep the implementations in `lax_reference.py` still and just have the translation rule registrations here.

But I don't feel strongly about it.

Reply:
Now, primitive implementations (= an alternative backend) are defined separately from a `lax` API clone (=for testing purposes). What I like about this is that it clarifies intent and enforces the API to be defined in terms of primitives, not the other way around.